lib/loader/loader.js

```diff
@@ -217,6 +217,11 @@ var loader, define, requireModule, require, requirejs;
   };
 
   define = function(name, deps, callback) {
+    var module = registry[name];
+    if (module && module.state !== 'new' && module.state !== 'errored') {
```

Comment:
This condition looks odd to me. 

IMO, we should only allow clobbering if the module state is `new`. Reason being, the module isn't required by anyone so it is fine to update its definition.  For any other state _after_ `new` we should _not_ allow clobbering. 

With the `errored` state, the module has already gone reification and its deps have caused an exception state to occur which is why the parent module is put in `errored` state. The current behavior of the loader is that if the same module is required again, it will throw an exception. 

With this condition, if the module was redefined (such that it doesn't it will not throw an exception now) between the first require (where an exception) is thrown and the second require (it gives a result), we are essentially magically changing the expected outcome of reification. IMO, we should not be doing. Once a module is `reified` its behavior should be locked. 

If we want to change the behavior then we should allow even changing it on `finalized` as well. Though IMO that is wrong.